src/core/lib/surface/server.c

```diff
@@ -1402,15 +1352,13 @@ static grpc_call_error queue_call_request(grpc_exec_ctx *exec_ctx,
       rm = &rc->data.registered.registered_method->request_matcher;
       break;
   }
-  server->requested_calls_per_cq[cq_idx][request_id] = *rc;
-  gpr_free(rc);
-  if (gpr_stack_lockfree_push(rm->requests_per_cq[cq_idx], request_id)) {
+  if (gpr_locked_mpscq_push(&rm->requests_per_cq[cq_idx], &rc->request_link)) {
```

Comment:
Looking at the mpsc implementation, this may return true even if there are some nodes being popped (for example, in the mpscq_pop function, the queue->stub is pushed even when there are some nodes left in the queue. Not sure if that is a problem though (seems like it won't be).. but something I thought I should bring it up.